### PR TITLE
CRI: add support for dns cluster first policy

### LINKED ISSUE #43352: The CRI dockershim doesn't support ClusterFirstWithHostNet
**Kubernetes version** (use `kubectl version`):

HEAD (~v1.6.0-beta.4)

**Environment**: any
**What happened**:

I want to use ClusterFirstWithHostNet which was implemented in https://github.com/kubernetes/kubernetes/pull/29378, but only dockertools (legacy/non-CRI) was updated to support this feature.

The CRI dockershim should be modified to support this rewriting the resolv.conf file of Pods with HostNetwork=true as well.

This is a bug in v1.6, I hope it can be prioritized to be fixed in v1.6.1, since it's blocking me from using CRI. Since CRI is enabled by default, it's expected that it should have feature parity with the earlier implementation.

@kubernetes/sig-node-bugs @freehan @yujuhong @feiskyer @vefimova @bowei @thockin

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

PR #29378 introduces ClusterFirstWithHostNet policy but only dockertools was updated to support the feature. 

This PR updates kuberuntime to support it for all runtimes.

**Which issue this PR fixes** 

fixes #43352

**Special notes for your reviewer**:

Candidate for v1.6.

**Release note**:

```release-note
NONE
```

cc @thockin @luxas @vefimova @Random-Liu

### CODE REVIEW

**Comment by luxas** [comment]
@feiskyer Thanks!! I'll test and verify in a minute :)

**Comment by luxas** [comment]
@feiskyer @thockin This worked for me and solved my issue :+1:, thanks!

I think this is a v1.6.0 candidate then, given the super fast response by @feiskyer 
I said v1.6.1 since I expected it would take some days to get the patch and didn't want to delay the release, but given this is just a merge away, this can be merged now

**Comment by feiskyer** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

**Comment by luxas** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

**Comment by feiskyer** [comment]
@luxas Added a new commit to fix e2e tests. The tests failed because docker v1.12 doesn't support setting dns options togather with host network. But docker v1.13+ does support this.

**Comment by ncdc** [comment]
cc @smarterclayton @derekwaynecarr @knobunc @dcbw @danwinship @eparis

**Comment by luxas** [comment]
I can confirm that this PR fixes the problem. I hope this will be prioritized and merged into v1.6

**Review by vefimova — APPROVED** [review]

File: pkg/kubelet/dockershim/docker_sandbox.go [hunk 1]
```diff
@@ -92,6 +92,25 @@ func (ds *dockerService) RunPodSandbox(config *runtimeapi.PodSandboxConfig) (str
...
+	// file is shared by all containers of the same pod, and needs to be modified
+	// only once per pod.
+	if dnsConfig := config.GetDnsConfig(); dnsConfig != nil {
+		containerInfo, err := ds.client.InspectContainer(createResp.ID)
+		if err != nil {
+			return createResp.ID, fmt.Errorf("failed to inspect sandbox container for pod %q: %v", config.Metadata.Name, err)
+		}
+
+		if err := dockertools.RewriteResolvFile(containerInfo.ResolvConfPath, dnsConfig.Servers, dnsConfig.Searches, len(dnsConfig.Options) > 0); err != nil {
+			return createResp.ID, fmt.Errorf("rewrite resolf.conf faield for pod %q: %v", config.Metadata.Name, err)
```
Comment (thockin) [inline pkg/kubelet/dockershim/docker_sandbox.go]: s/resolf/resolv

**Comment by thockin** [comment]
Thanks!

On Mon, Mar 20, 2017 at 10:39 AM, Victoria Efimova <notifications@github.com

File: pkg/kubelet/dockershim/docker_sandbox.go (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/dockershim/docker_sandbox.go]: Ack. Will open another pr to fix this.

**Pull request merged**